ompi/mpi/c/finalized.c

```diff
@@ -48,17 +51,28 @@ int MPI_Finalized(int *flag)
                MPI_Finalize) or not */
 
             if (ompi_mpi_initialized && !ompi_mpi_finalized) {
+                opal_mutex_unlock(&ompi_mpi_bootstrap_mutex);
                 return OMPI_ERRHANDLER_INVOKE(MPI_COMM_WORLD, MPI_ERR_ARG,
                                               FUNC_NAME);
             } else {
+                opal_mutex_unlock(&ompi_mpi_bootstrap_mutex);
                 return OMPI_ERRHANDLER_INVOKE(null, MPI_ERR_ARG,
                                               FUNC_NAME);
             }
         }
     }
 
-    /* Pretty simple */
+    /* If some process has *started* FINALIZE but not yet completed
+       it, then be a little social and wait here until it is
+       completed. */
+    while (ompi_mpi_finalize_started && !ompi_mpi_finalized) {
+        opal_mutex_unlock(&ompi_mpi_bootstrap_mutex);
+        usleep(1);
+        opal_mutex_lock(&ompi_mpi_bootstrap_mutex);
+    }
 
```

Comment:
The result of this loop is basically a continuous loop doing a "lock + unlock"every 1us. As you are using a mutex, why don't you rely on the trigger of the mutex by the thread that succeeded to ensure the sequentiality of this code?